feat: Bedrock Agentcore MCP (clean)

w601sxs opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
Fixes
- please ignore prev. pull request https://github.com/awslabs/mcp/pull/1196
- cleaned precommit hooks with all tests passing locally

## Summary
Adding MCP server for Agentcore including support for:

transoforming existing agentcode to agetncore compatible code
Creating and managing gateways
Creating memory resources and using with agents
Using Agentcore Identity

### Changes

> New MCP server for the Bedrock Agentcore services

### User experience

Amazon Bedrock AgentCore MCP server




Installation

One-click install:

Q CLI integration                                               Kiro:


Working with MCP inspector: 



Example intents covered

* “What tools do you have access to?”


* “Convert this code to agentcore”/ “Transform to agentcore” / “ Make this agentcore compatible” 


* “Troubleshoot errors”/“do you see any errors on cloudwatch?”


* “What agents can I invoke?” / “can you test agentv3”/ “say hello to strandsv4 agent”


* “Create a dynamodb gateway” (downloads smithy model directly and deploys)
* “what tools are available in this gateway?”



* “Can you find me some good examples for...“ / ”I need an example for hosting strands on runtime“ / .... (finds curated examples from our samples github)


* deploy to runtime, but a version already exists


* recovery from api param errors (runtime)


* uses codebuild/auto-role creation for runtime


* “What agents can I invoke?” / “Is strandsagentv6 ready to invoke?” 


* deploy to runtime with oauth


* Invoke runtime with oauth / “can you generate a bearer token”


* list memory resources I can use


* create memory resource / use this memory in my agent code



## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [ X ] I have reviewed the [contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [ X ] I have performed a self-review of this change
* [ X ] Changes have been tested
* [ X ] Changes are documented

Is this a breaking change? (Y/N)
N
**RFC issue number**:
1079

Checklist:

* [ X ] Migration process documented
* [ X ] Implement warnings (if it can live side by side)

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE).

jamesward commented on src/amazon-bedrock-agentcore-mcp-server/README.md:
```diff
@@ -0,0 +1,410 @@
+1. [Sign-Up for an AWS account](https://aws.amazon.com/free/?trk=78b916d7-7c94-4cab-98d9-0ce5e648dd5f&sc_channel=ps&ef_id=Cj0KCQjwxJvBBhDuARIsAGUgNfjOZq8r2bH2OfcYfYTht5v5I1Bn0lBKiI2Ii71A8Gk39ZU5cwMLPkcaAo_CEALw_wcB:G:s&s_kwcid=AL!4422!3!432339156162!e!!g!!aws%20sign%20up!9572385111!102212379327&gad_campaignid=9572385111&gbraid=0AAAAADjHtp99c5A9DUyUaUQVhVEoi8of3&gclid=Cj0KCQjwxJvBBhDuARIsAGUgNfjOZ ...
+2. [Set up Amazon Bedrock Agent Core](https://docs.aws.amazon.com/bedrock/latest/userguide/agents.html) access in your AWS account
+3. Install `uv` from [Astral](https://docs.astral.sh/uv/getting-started/installation/) or the [GitHub README](https://github.com/astral-sh/uv#installation)
+4. Install Python using `uv python install 3.10`
```
Since using `uv` there should be no need to install `python`

w601sxs replied:
removed.

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/runtime.py:
```diff
@@ -0,0 +1,2606 @@
+        return generate_basic_agentcore_code(original_code, analysis, options)
+
+
+def generate_strands_agentcore_code(original_code: str, options: Dict[str, bool]) -> str:
```
I tried both, and depending on the client this did/did not work. I can add some more functionality to return to client instead of modifying code

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/server.py:
```diff
@@ -0,0 +1,434 @@
+
+
+@mcp.tool()
+async def server_info() -> str:
```
Is server.py required? I found it in other folders. Can I remove this entirely?

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/utils.py:
```diff
@@ -0,0 +1,2859 @@
+## ============================================================================
+
+
+def register_oauth_tools(mcp: FastMCP):
```
moved to identity!

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/server.py:
```diff
@@ -0,0 +1,434 @@
+
+# === MODULE IMPORTS ===
+# Import all modular components with comprehensive error handling
+try:
```
changed this in the new commits

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/identity.py:
```diff
@@ -0,0 +1,899 @@
+
+import json
+import os
+import subprocess
```
we have considered this

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/identity.py:
```diff
@@ -0,0 +1,450 @@
+Secure: Security: API key is securely stored and encrypted by AWS AgentCore Identity service."""
+
+                except Exception as create_error:
+                    return f"""X Failed to Create Credential Provider
```
yes changed throughout

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/identity.py:
```diff
@@ -0,0 +1,450 @@
+- Reuse: Reusable across multiple agents"""
+
+                    # Format results
+                    result_parts = []
```
changed

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/runtime.py:
```diff
@@ -0,0 +1,2606 @@
+        # Check if uv is available for package management
+        uv_available = False
+        try:
+            result = subprocess.run(['which', 'uv'], capture_output=True, text=True)
```
considered this

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/runtime.py:
```diff
@@ -0,0 +1,2606 @@
+    code_parts.append('')
+    code_parts.append('from bedrock_agentcore import BedrockAgentCoreApp')
+    code_parts.append('from strands import Agent')
+    code_parts.append('')
```
considered this

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/gateway.py:
```diff
@@ -0,0 +1,1876 @@
+
+        # Step 3: Get contents of service directory to find version
+        service_api_url = service_dir['url']
+        response = requests.get(service_api_url)
```
considered this

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/gateway.py:
```diff
@@ -0,0 +1,1876 @@
+
+        # Step 4: Get contents of version directory to find service.json
+        version_api_url = version_dir['url']
+        response = requests.get(version_api_url)
```
considered this

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/runtime.py:
```diff
@@ -0,0 +1,2606 @@
+
+            # Check for uv
+            try:
+                result = subprocess.run(['which', 'uv'], capture_output=True, text=True)
```
considered this

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/runtime.py:
```diff
@@ -0,0 +1,2606 @@
+            config_exists, config_dir = find_agent_config_directory(agent_name)
+
+            if config_exists:
+                result = subprocess.run(
+                    cmd, capture_output=True, text=True, timeout=30, cwd=config_dir
```
considered this

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/runtime.py:
```diff
@@ -0,0 +1,2606 @@
+                    ):
+                        discovered_agents.append(agent_info)
+
+                except Exception:
+                    # Skip invalid config files
```
considered this

w601sxs commented:
> @w601sxs Overall comments:
> 
> 1. consider defining complex error messages separated from the implementation code
> 2. Forcing your files to <600-700 LOCs will help to think through how to modularize it
> 3. I think it needs a bit more strategic thought/mental model on how it will be used. E.g. is this for the vibe coder who wants to develop agents? or is it for agents themselves (e.g. to use memory?),
> 4. We should consider the current AWS API coverage already available in AWS API MCP
> 5. consider strategically what parts should be done in the server vs. relying on the LLM in the MCP client itself

@theagenticguy thank you for your comments.

1. I have refactored what used to be returned as errors and warnings to raising specific errors (Value, OS, IO, MCP tool error) and separated out what should be raised vs returned as context to the client

2. This is still in progress, our natural separate lines are in the services themselves (like runtime, gateway, identity, etc) which we are ok maintaining. We expect individual sub-service specific owners to choose to modularize or keep it the same way. 

3. This is mainly for the vibe coder persona. this does not preclude needing to deploy and test agents, as we think this is good to have as part of making the devex easier

4. AWS API, and AWS docs should be used alongside. We will cover this in an upcoming blog. But, we also want to have the ability to use this independently

5.  Some recent commits changed the way we do this - for example, we experimented with transforming code, changed instructions, and follow a hybrid approach now where an automated transformation is attempted but the client is encouraged to test/check again before proceeding with launch in runtime.py


Most commits are trying to fix tests that failed, which was difficult since most of those were API calls that needed to be mocked in the absence of auth.

theagenticguy commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/memory.py:
```diff
@@ -0,0 +1,1103 @@
+    """Register memory management tools."""
+
+    @mcp.tool()
+    async def memory_save_conversation(
```
How will an agent use this tool to save a conversation turn?

theagenticguy commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/memory.py:
```diff
@@ -0,0 +1,1103 @@
+
+        Stores user input and agent response as a conversation turn in the specified memory.
+        """
+        if not SDK_AVAILABLE:  # pragma: no cover
```
this if statement seems to be consistent across the board. should it be moved earlier in the server initialization instead? or shouldn't this SDK always be available anyway based on pyproject.toml?

theagenticguy commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/memory.py:
```diff
@@ -0,0 +1,1103 @@
+- Session ID: {session_id}""")
+
+    @mcp.tool()
+    async def memory_create_event(
```
same as above.

theagenticguy commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/memory.py:
```diff
@@ -0,0 +1,1103 @@
+
+        try:
+            # Import memory-related classes
+            from bedrock_agentcore.memory import MemoryClient, MemoryControlPlaneClient
```
move imports to the top of the files

w601sxs replied:
changed

theagenticguy commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/memory.py:
```diff
@@ -0,0 +1,1103 @@
+                    # Wait for memory to become active
+                    create_steps.append(' Waiting for memory to become active...')
+
+                    max_wait_time = 180  # 3 minutes
+                    wait_start = time.time()
+
+                    while time.time() - wait_start < max_wait_time:
```
Consider:

```max_wait_time = 180  # 3 minutes
deadline = time.monotonic() + max_wait_time

while time.monotonic() < deadline:```

theagenticguy commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/memory.py:
```diff
@@ -0,0 +1,1103 @@
+
+                    # Optional: Integrate with agent file if provided
+                    integration_result = ''
+                    if agent_file:
```
I think we should avoid editing code files this way as we discussed offline

w601sxs replied:
removed

theagenticguy commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/gateway.py:
```diff
@@ -0,0 +1,1903 @@
+Issue: {len(final_remaining_ids)} targets could not be deleted after {max_deletion_attempts} attempts with throttling protection
+
+## Deletion Steps Attempted:
+{chr(10).join(f'- {step}' for step in deletion_steps)}
```
Consider:

```python
print(f"""This is a multiline thing

- {'\n - '.join(steps)}
""")
```

w601sxs replied:
changed but \n is not allowed in f string expression is what it says

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/.amazonq/mcp.json:
removed .amazonq folder

w601sxs commented on src/amazon-bedrock-agentcore-mcp-server/awslabs/amazon_bedrock_agentcore_mcp_server/gateway.py:
```diff
@@ -0,0 +1,1903 @@
+Issue: {len(final_remaining_ids)} targets could not be deleted after {max_deletion_attempts} attempts with throttling protection
+
+## Deletion Steps Attempted:
+{chr(10).join(f'- {step}' for step in deletion_steps)}
```
changed
